dsp/modules/googlevertexai.py

```diff
@@ -0,0 +1,174 @@
+
+try:
+    import vertexai  # type: ignore[import-untyped]
+    from vertexai.language_models import CodeGenerationModel, TextGenerationModel
+    from vertexai.preview.generative_models import GenerativeModel
+except ImportError:
+    pass
+
+
+def backoff_hdlr(details):
+    """Handler from https://pypi.org/project/backoff/"""
+    print(f"Backing off {details['wait']:0.1f} seconds after {details['tries']} tries "
+          f"calling function {details['target']} with kwargs "
+          f"{details['kwargs']}")
+
+
+def giveup_hdlr(details):
+    """wrapper function that decides when to give up on retry"""
+    if "rate limits" in details.message:
+        return False
+    return True
+
+class GoogleVertexAI(LM):
+    """Wrapper around GoogleVertexAI's API.
+
+    Currently supported models include `gemini-pro-1.0`.
+    """
+
+    def __init__(
+        self, model_name: str = "text-bison@002", **kwargs,
+    ):
+        """
+        Parameters
+        ----------
+        model : str
+            Which pre-trained model from Google to use?
+            Choices are [`text-bison@002`]
+        **kwargs: dict
+            Additional arguments to pass to the API provider.
+        """
+        super().__init__(model_name)
+        self._is_gemini = "gemini" in model_name
+        self._init_vertexai(kwargs)
+        if "code" in model_name:
+            model_cls = CodeGenerationModel
+            self.available_args = {
+                'suffix',
+                'max_output_tokens',
+                'temperature',
+                'stop_sequences',
+                'candidate_count',
+            }
+        elif "gemini" in model_name:
+            model_cls = GenerativeModel
+            self.available_args = {
+                'max_output_tokens',
+                'temperature',
+                'top_k',
+                'top_p',
+                'stop_sequences',
+                'candidate_count',
+            }
+        elif 'text' in model_name:
+            model_cls = TextGenerationModel
+            self.available_args = {
+                'max_output_tokens',
+                'temperature',
+                'top_k',
+                'top_p',
+                'stop_sequences',
+                'candidate_count',
+            }
+        else:
+            raise PydanticCustomError(
+                'model',
+                'model name is not valid, got "{model_name}"',
+                dict(wrong_value=model_name),
+            )
+        if self._is_gemini:
+            self.client = model_cls(model_name=model_name, safety_settings=kwargs.get('safety_settings')) # pylint: disable=unexpected-keyword-arg,no-value-for-parameter
+        else:
+            self.client = model_cls.from_pretrained(model_name)
+        self.provider = "googlevertexai"
+        self.kwargs = {
+            **self.kwargs,
+            "temperature": 0.7,
+            "max_output_tokens": 1024,
+            "top_p": 1.0,
+            "top_k": 1,
```

Comment:
why are these kwargs available for non-gemini models if they are set to defaults for non-gemini models?